src/uct/ib/ud/base/ud_ep.c

```diff
@@ -1757,14 +1836,36 @@ void uct_ud_ep_disconnect(uct_ep_h tl_ep)
      * to not invoke user callbacks for the endpoint which was destroyed */
     uct_ud_ep_purge(ep, UCS_ERR_CANCELED);
 
+    if (!(ep->flags & UCT_UD_EP_FLAG_CONNECT_TO_EP) &&
+        (ep->flags & UCT_UD_EP_FLAG_TX)) {
+        /* Remove the endpoint from the connection matching context if it
+         * exists there */
+        uct_ud_iface_cep_remove_ep(iface, ep);
+        ep->flags &= ~UCT_UD_EP_FLAG_TX;
+        uct_ud_ep_ctl_op_del(ep, UCT_UD_EP_OP_CREQ);
+
+        if (uct_ud_ep_is_connected(ep)) {
+            uct_ud_ep_ctl_op_add(iface, ep, UCT_UD_EP_OP_FIN);
+        }
+
+        if (ep->flags & UCT_UD_EP_FLAG_RX) {
+            uct_ud_iface_cep_insert_ready_ep(iface, ep);
```

Comment:
also here, if we destroyed ep form API, how could it be ready for matching with new ep from remote side?

Reply:
if there is RX flag, then it means that it is able to receive data from a peer even if local user closed this EP.
but keeping this EP in CEP is needed for consistency for the following rule: "UD EP is in connection matching unexpected queue if it has only RX capability, expected queue - if it has only TX capability or both RX/TX capabilities"